File: model/advdiff.f
      double precision function advdiff( qt )
c
c   calculate divergence from advective-diffusive balance for 
c   condensate in a model layer
c
c   all units are cgs
c
c   input parameters passed through argument list:
c
c     qt              total mixing ratio of condensate + vapor (g/g)
c
c   input parameters passed through common vfall_block:
c
c     ad_qbelow       total mixing ratio of vapor in underlying layer (g/g)
c     ad_qvs          saturation mixing ratio (g/g)
c     ad_mixl         convective mixing length (cm)
c     ad_dz           layer thickness (cm) 
c     ad_rainf        rain efficiency factor 
c
c   output parameters passed through common vfall_block:
c
c     ad_qc           mixing ratio of condensed condensate (g/g)
c   
c
c   A. Ackerman Feb-2000
c
      implicit none


c   Declare common storage 

      double precision ad_qbelow, ad_qvs, ad_mixl, ad_dz, ad_rainf

      common / advdiff_block /
     $  ad_qbelow, ad_qvs, ad_mixl, ad_dz, ad_rainf


c   Declare local storage

      double precision qt, ad_qc


c   All vapor in excess of saturation condenses (supsat=0)
      ad_qc = max( 0., qt - ad_qvs )

c   Difference from advective-diffusive balance 
      advdiff = 
     $   ad_qbelow*exp( - ad_rainf*ad_qc*ad_dz / ( qt*ad_mixl ) ) - qt

      return
      end

File: python_v2/advdiff.py
import math

def advdiff(qt: float, ad_qbelow: float, ad_qvs: float,
            ad_mixl: float, ad_dz: float, ad_rainf: float) -> tuple[float, float]:
    """
    Calculate divergence from advective-diffusive balance for condensate in a model layer.

    Parameters
    ----------
    qt : float
        Total mixing ratio of condensate + vapor (g/g).
    ad_qbelow : float
        Total mixing ratio of vapor in underlying layer (g/g).
    ad_qvs : float
        Saturation mixing ratio (g/g).
    ad_mixl : float
        Convective mixing length (cm).
    ad_dz : float
        Layer thickness (cm).
    ad_rainf : float
        Rain efficiency factor.

    Returns
    -------
    advdiff_val : float
        Difference from advective-diffusive balance.
    ad_qc : float
        Mixing ratio of condensed condensate (g/g).
    """
    # All vapor in excess of saturation condenses
    ad_qc = max(0.0, qt - ad_qvs)

    # Difference from advective-diffusive balance
    advdiff_val = ad_qbelow * math.exp(-ad_rainf * ad_qc * ad_dz / (qt * ad_mixl)) - qt

    return advdiff_val, ad_qc
